File: module-infra/src/main/java/maple/expectation/infrastructure/persistence/repository/RedisRefreshTokenRepositoryImpl.java
package maple.expectation.infrastructure.persistence.repository;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.Duration;
import java.util.List;
import java.util.Set;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.domain.RefreshToken;
import maple.expectation.domain.repository.RedisRefreshTokenRepository;
import maple.expectation.infrastructure.executor.LogicExecutor;
import maple.expectation.infrastructure.executor.TaskContext;
import org.jspecify.annotations.Nullable;
import org.redisson.api.RBucket;
import org.redisson.api.RScript;
import org.redisson.api.RSet;
import org.redisson.api.RedissonClient;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Repository;

/**
 * Redis 기반 Refresh Token 저장소 구현체 (Issue #279)
 *
 * <p>저장 구조:
 *
 * <ul>
 *   <li>Token: refresh:{refreshTokenId} → JSON (String)
 *   <li>Family Index: refresh:family:{familyId} → Set&lt;refreshTokenId&gt;
 *   <li>Session Index: refresh:session:{sessionId} → Set&lt;refreshTokenId&gt;
 * </ul>
 *
 * <p>TTL 정책:
 *
 * <ul>
 *   <li>Token TTL: 7일 (auth.refresh-token.expiration)
 *   <li>Family/Session Index TTL: 7일 (토큰과 동일)
 * </ul>
 */
@Slf4j
@Repository
public class RedisRefreshTokenRepositoryImpl implements RedisRefreshTokenRepository {

  private static final String KEY_PREFIX = "refresh:";
  private static final String FAMILY_KEY_PREFIX = "refresh:family:";
  private static final String SESSION_KEY_PREFIX = "refresh:session:";

  /**
   * ADR-082: Atomic Lua script for check-and-mark operation.
   *
   * <p>Uses Redis cjson library for proper JSON parsing (not pattern matching).
   *
   * <p>Returns:
   *
   * <ul>
   *   <li>nil: Token not found
   *   <li>"ALREADY_USED": Token reuse detected (used=true)
   *   <li>updated JSON: Token successfully marked as used
   * </ul>
   */
  private static final String ATOMIC_CHECK_AND_MARK_LUA =
      """
      local tokenJson = redis.call('GET', KEYS[1])
      if tokenJson == false then
          return nil
      end

      local token = cjson.decode(tokenJson)

      if token.used == true then
          return 'ALREADY_USED'
      end

      token.used = true
      local newJson = cjson.encode(token)

      if ARGV[1] ~= '0' then
          redis.call('PSETEX', KEYS[1], ARGV[1], newJson)
      else
          redis.call('SET', KEYS[1], newJson)
      end

      return newJson
      """;

  private final RedissonClient redissonClient;
  private final ObjectMapper objectMapper;
  private final LogicExecutor executor;
  private final long refreshTokenTtlSeconds;

  public RedisRefreshTokenRepositoryImpl(
      RedissonClient redissonClient,
      ObjectMapper objectMapper,
      LogicExecutor executor,
      @Value("${auth.refresh-token.expiration}") long refreshTokenTtlSeconds) {
    this.redissonClient = redissonClient;
    this.objectMapper = objectMapper;
    this.executor = executor;
    this.refreshTokenTtlSeconds = refreshTokenTtlSeconds;
  }

  /**
   * Refresh Token 저장
   *
   * @param token 저장할 Refresh Token
   */
  public void save(RefreshToken token) {
    executor.executeVoidJava(
        () -> {
          String key = buildTokenKey(token.refreshTokenId());
          String json = serializeToken(token);

          // 1. 토큰 저장
          RBucket<String> bucket = redissonClient.getBucket(key);
          bucket.set(json, Duration.ofSeconds(refreshTokenTtlSeconds));

          // 2. Family Index에 추가 (탈취 감지 시 일괄 삭제용)
          String familyKey = buildFamilyKey(token.familyId());
          RSet<String> familySet = redissonClient.getSet(familyKey);
          familySet.add(token.refreshTokenId());
          familySet.expire(Duration.ofSeconds(refreshTokenTtlSeconds));

          // 3. Session Index에 추가 (로그아웃 시 일괄 삭제용)
          String sessionKey = buildSessionKey(token.sessionId());
          RSet<String> sessionSet = redissonClient.getSet(sessionKey);
          sessionSet.add(token.refreshTokenId());
          sessionSet.expire(Duration.ofSeconds(refreshTokenTtlSeconds));
        },
        TaskContext.of("RefreshToken", "Save", token.refreshTokenId()));

    log.debug(
        "RefreshToken saved: tokenId={}, familyId={}", token.refreshTokenId(), token.familyId());
  }

  @Override
  public @Nullable RefreshToken findById(String refreshTokenId) {
    return executor.executeOrDefault(
        () -> doFindById(refreshTokenId),
        null,
        TaskContext.of("RefreshToken", "FindById", refreshTokenId));
  }

  private @Nullable RefreshToken doFindById(String refreshTokenId) {
    String key = buildTokenKey(refreshTokenId);
    RBucket<String> bucket = redissonClient.getBucket(key);
    String json = bucket.get();

    if (json == null) {
      return null;
    }

    return deserializeToken(json);
  }

  /**
   * Refresh Token 사용 처리 (Token Rotation)
   *
   * <p>기존 토큰의 used 필드를 true로 설정하여 재사용 감지 가능하게 함
   *
   * @param refreshTokenId Refresh Token ID
   */
  public void markAsUsed(String refreshTokenId) {
    executor.executeVoidJava(
        () -> {
          String key = buildTokenKey(refreshTokenId);
          RBucket<String> bucket = redissonClient.getBucket(key);
          String json = bucket.get();

          if (json != null) {
            RefreshToken token = deserializeToken(json);
            RefreshToken usedToken = token.markAsUsed();
            bucket.set(
                serializeToken(usedToken),
                bucket.remainTimeToLive(),
                java.util.concurrent.TimeUnit.MILLISECONDS);
          }
        },
        TaskContext.of("RefreshToken", "MarkAsUsed", refreshTokenId));

    log.debug("RefreshToken marked as used: tokenId={}", refreshTokenId);
  }

  /**
   * Atomic Check-and-Mark: 토큰 사용 상태 확인 후 마크 (P1 Race Condition Fix)
   *
   * <p>Redis Lua script로 원자적으로 수행하여 TOCTOU 취약점 방지:
   *
   * <ul>
   *   <li>토큰이 존재하지 않으면 null 반환
   *   <li>이미 used=true이면 null 반환 (재사용 감지)
   *   <li>used=false이면 used=true로 변경 후 토큰 반환
   * </ul>
   *
   * @param refreshTokenId Refresh Token ID
   * @return 마크된 RefreshToken (이미 사용되었거나 존재하지 않으면 null)
   */
  @Override
  public @Nullable RefreshToken checkAndMarkAsUsed(String refreshTokenId) {
    return executor.executeOrDefault(
        () -> doCheckAndMarkAsUsed(refreshTokenId),
        null,
        TaskContext.of("RefreshToken", "CheckAndMark", refreshTokenId));
  }

  /**
   * ADR-082: Atomic check-and-mark using Lua script (P0 Fix).
   *
   * <p>Replaces non-atomic get-then-set with atomic Lua script execution.
   *
   * <p>Uses Redis cjson.decode() for proper JSON parsing instead of pattern matching. Previous
   * attempt used string.match(tokenJson, '"used":(true|false)') which failed because Lua pattern
   * matching doesn't treat | as alternation.
   *
   * <p>Return values from Lua script:
   *
   * <ul>
   *   <li>nil: Token not found → null
   *   <li>"ALREADY_USED": Token reuse detected → null
   *   <li>updated JSON: Success → RefreshToken
   * </ul>
   *
   * @param refreshTokenId Refresh Token ID
   * @return Marked RefreshToken if successful, null if not found or already used
   */
  private @Nullable RefreshToken doCheckAndMarkAsUsed(String refreshTokenId) {
    String key = buildTokenKey(refreshTokenId);
    RBucket<String> bucket = redissonClient.getBucket(key);

    // Read current TTL to preserve it after update
    long remainingTtl =
        executor.executeOrDefault(
            () -> bucket.remainTimeToLive(),
            0L,
            TaskContext.of("RefreshToken", "GetTTL", refreshTokenId));

    if (remainingTtl < 0) {
      // Key doesn't exist (remainTimeToLive returns -2 for non-existent keys)
      log.debug("Token not found in Redis: key={}", key);
      return null;
    }

    // Execute atomic Lua script
    RScript script = redissonClient.getScript();
    String result =
        script.eval(
            RScript.Mode.READ_WRITE,
            ATOMIC_CHECK_AND_MARK_LUA,
            RScript.ReturnType.VALUE,
            List.of(key),
            String.valueOf(remainingTtl > 0 ? remainingTtl : 0));

    // Handle Lua script return values
    if (result == null) {
      log.debug("Token not found in Redis (Lua): key={}", key);
      return null;
    }

    if ("ALREADY_USED".equals(result)) {
      log.warn("Token reuse detected! Token is already marked as used: key={}", key);
      return null;
    }

    // Success: token was marked as used
    log.debug("Token marked as used atomically: key={}", key);
    return deserializeToken(result);
  }

  /**
   * Family 전체 무효화 (탈취 감지 시)
   *
   * @param familyId Token Family ID
   */
  public void deleteByFamilyId(String familyId) {
    executor.executeVoidJava(
        () -> {
          String familyKey = buildFamilyKey(familyId);
          RSet<String> familySet = redissonClient.getSet(familyKey);
          Set<String> tokenIds = familySet.readAll();

          // Family에 속한 모든 토큰 삭제
          for (String tokenId : tokenIds) {
            String tokenKey = buildTokenKey(tokenId);
            redissonClient.getBucket(tokenKey).delete();
          }

          // Family Index 삭제
          familySet.delete();
        },
        TaskContext.of("RefreshToken", "DeleteByFamily", familyId));

    log.warn("Token family invalidated (possible token theft): familyId={}", familyId);
  }

  /**
   * 세션의 모든 Refresh Token 삭제 (로그아웃 시)
   *
   * @param sessionId 세션 ID
   */
  public void deleteBySessionId(String sessionId) {
    executor.executeVoidJava(
        () -> {
          String sessionKey = buildSessionKey(sessionId);
          RSet<String> sessionSet = redissonClient.getSet(sessionKey);
          Set<String> tokenIds = sessionSet.readAll();

          // 세션에 연결된 모든 토큰 삭제
          for (String tokenId : tokenIds) {
            String tokenKey = buildTokenKey(tokenId);
            RBucket<String> bucket = redissonClient.getBucket(tokenKey);
            String json = bucket.get();

            if (json != null) {
              RefreshToken token = deserializeToken(json);
              // Family Index에서도 제거
              String familyKey = buildFamilyKey(token.familyId());
              redissonClient.getSet(familyKey).remove(tokenId);
            }

            bucket.delete();
          }

          // Session Index 삭제
          sessionSet.delete();
        },
        TaskContext.of("RefreshToken", "DeleteBySession", sessionId));

    log.debug("RefreshTokens deleted for session: sessionId={}", sessionId);
  }

  /**
   * 단일 Refresh Token 삭제
   *
   * @param refreshTokenId Refresh Token ID
   */
  public void deleteById(String refreshTokenId) {
    executor.executeVoidJava(
        () -> {
          String tokenKey = buildTokenKey(refreshTokenId);
          redissonClient.getBucket(tokenKey).delete();
        },
        TaskContext.of("RefreshToken", "DeleteById", refreshTokenId));
  }

  private String buildTokenKey(String refreshTokenId) {
    return KEY_PREFIX + refreshTokenId;
  }

  private String buildFamilyKey(String familyId) {
    return FAMILY_KEY_PREFIX + familyId;
  }

  private String buildSessionKey(String sessionId) {
    return SESSION_KEY_PREFIX + sessionId;
  }

  private String serializeToken(RefreshToken token) {
    return executor.execute(
        () -> objectMapper.writeValueAsString(token),
        TaskContext.of("RefreshToken", "Serialize", token.refreshTokenId()));
  }

  private RefreshToken deserializeToken(String json) {
    return executor.execute(
        () -> objectMapper.readValue(json, RefreshToken.class),
        TaskContext.of(
            "RefreshToken", "Deserialize", json.length() > 30 ? json.substring(0, 30) : json));
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/persistence/repository/RedisRefreshTokenRepositoryImpl.kt
package maple.expectation.infrastructure.persistence.repository

import com.fasterxml.jackson.databind.ObjectMapper
import maple.expectation.core.domain.auth.RefreshToken
import maple.expectation.domain.repository.RedisRefreshTokenRepository as DomainRedisRefreshTokenRepository
import maple.expectation.infrastructure.executor.LogicExecutor
import maple.expectation.infrastructure.executor.TaskContext
import org.jspecify.annotations.Nullable
import org.redisson.api.RBucket
import org.redisson.api.RScript
import org.redisson.api.RSet
import org.redisson.api.RedissonClient
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Repository
import java.time.Duration
import java.util.List

@Repository
open class RedisRefreshTokenRepositoryImpl(
    private val redissonClient: RedissonClient,
    private val objectMapper: ObjectMapper,
    private val executor: LogicExecutor,
    @Value("\${auth.refresh-token.expiration}") private val refreshTokenTtlSeconds: Long,
) : DomainRedisRefreshTokenRepository {

    private companion object {
        const val KEY_PREFIX = "refresh:"
        const val FAMILY_KEY_PREFIX = "refresh:family:"
        const val SESSION_KEY_PREFIX = "refresh:session:"

        val ATOMIC_CHECK_AND_MARK_LUA = """
            local tokenJson = redis.call('GET', KEYS[1])
            if tokenJson == false then
                return nil
            end
            local token = cjson.decode(tokenJson)
            if token.used == true then
                return 'ALREADY_USED'
            end
            token.used = true
            local newJson = cjson.encode(token)
            if ARGV[1] ~= '0' then
                redis.call('PSETEX', KEYS[1], ARGV[1], newJson)
            else
                redis.call('SET', KEYS[1], newJson)
            end
            return newJson
        """.trimIndent()
    }

    override fun save(token: RefreshToken) {
        executor.executeVoidJava(
            Runnable {
                val key = buildTokenKey(token.refreshTokenId())
                val json = serializeToken(token)

                val bucket = redissonClient.getBucket<String>(key)
                bucket.set(json, Duration.ofSeconds(refreshTokenTtlSeconds))

                val familyKey = buildFamilyKey(token.familyId())
                val familySet = redissonClient.getSet<String>(familyKey)
                familySet.add(token.refreshTokenId())
                familySet.expire(Duration.ofSeconds(refreshTokenTtlSeconds))

                val sessionKey = buildSessionKey(token.sessionId())
                val sessionSet = redissonClient.getSet<String>(sessionKey)
                sessionSet.add(token.refreshTokenId())
                sessionSet.expire(Duration.ofSeconds(refreshTokenTtlSeconds))
            },
            TaskContext.of("RefreshToken", "Save", token.refreshTokenId()),
        )
    }

    override fun findById(refreshTokenId: String): RefreshToken? {
        return executor.executeOrDefault(
            { doFindById(refreshTokenId) },
            null,
            TaskContext.of("RefreshToken", "FindById", refreshTokenId),
        )
    }

    private fun doFindById(refreshTokenId: String): RefreshToken? {
        val key = buildTokenKey(refreshTokenId)
        val bucket = redissonClient.getBucket<String>(key)
        val json = bucket.get()
        return json?.let { deserializeToken(it) }
    }

    override fun markAsUsed(refreshTokenId: String) {
        executor.executeVoidJava(
            Runnable {
                val key = buildTokenKey(refreshTokenId)
                val bucket = redissonClient.getBucket<String>(key)
                val json = bucket.get()
                if (json != null) {
                    val token = deserializeToken(json)
                    val usedToken = token.markAsUsed()
                    bucket.set(
                        serializeToken(usedToken),
                        bucket.remainTimeToLive(),
                        java.util.concurrent.TimeUnit.MILLISECONDS,
                    )
                }
            },
            TaskContext.of("RefreshToken", "MarkAsUsed", refreshTokenId),
        )
    }

    override fun checkAndMarkAsUsed(refreshTokenId: String): RefreshToken? {
        return executor.executeOrDefault(
            { doCheckAndMarkAsUsed(refreshTokenId) },
            null,
            TaskContext.of("RefreshToken", "CheckAndMark", refreshTokenId),
        )
    }

    private fun doCheckAndMarkAsUsed(refreshTokenId: String): RefreshToken? {
        val key = buildTokenKey(refreshTokenId)
        val bucket = redissonClient.getBucket<String>(key)

        val remainingTtl =
            executor.executeOrDefault(
                { bucket.remainTimeToLive() },
                0L,
                TaskContext.of("RefreshToken", "GetTTL", refreshTokenId),
            )

        if (remainingTtl < 0) {
            return null
        }

        val script = redissonClient.script
        val result =
            script.eval(
                RScript.Mode.READ_WRITE,
                ATOMIC_CHECK_AND_MARK_LUA,
                RScript.ReturnType.VALUE,
                listOf(key),
                (if (remainingTtl > 0) remainingTtl else 0).toString(),
            ) as? String

        return when (result) {
            null -> null
            "ALREADY_USED" -> null
            else -> deserializeToken(result)
        }
    }

    override fun deleteByFamilyId(familyId: String) {
        executor.executeVoidJava(
            Runnable {
                val familyKey = buildFamilyKey(familyId)
                val familySet = redissonClient.getSet<String>(familyKey)
                val tokenIds = familySet.readAll()

                for (tokenId in tokenIds) {
                    val tokenKey = buildTokenKey(tokenId)
                    redissonClient.getBucket<Any>(tokenKey).delete()
                }
                familySet.delete()
            },
            TaskContext.of("RefreshToken", "DeleteByFamily", familyId),
        )
    }

    override fun deleteBySessionId(sessionId: String) {
        executor.executeVoidJava(
            Runnable {
                val sessionKey = buildSessionKey(sessionId)
                val sessionSet = redissonClient.getSet<String>(sessionKey)
                val tokenIds = sessionSet.readAll()

                for (tokenId in tokenIds) {
                    val tokenKey = buildTokenKey(tokenId)
                    val bucket = redissonClient.getBucket<String>(tokenKey)
                    val json = bucket.get()

                    if (json != null) {
                        val token = deserializeToken(json)
                        val familyKey = buildFamilyKey(token.familyId())
                        redissonClient.getSet<String>(familyKey).remove(tokenId)
                    }
                    bucket.delete()
                }
                sessionSet.delete()
            },
            TaskContext.of("RefreshToken", "DeleteBySession", sessionId),
        )
    }

    override fun deleteById(refreshTokenId: String) {
        executor.executeVoidJava(
            Runnable {
                val tokenKey = buildTokenKey(refreshTokenId)
                redissonClient.getBucket<Any>(tokenKey).delete()
            },
            TaskContext.of("RefreshToken", "DeleteById", refreshTokenId),
        )
    }

    private fun buildTokenKey(refreshTokenId: String): String = KEY_PREFIX + refreshTokenId

    private fun buildFamilyKey(familyId: String): String = FAMILY_KEY_PREFIX + familyId

    private fun buildSessionKey(sessionId: String): String = SESSION_KEY_PREFIX + sessionId

    private fun serializeToken(token: RefreshToken): String {
        return executor.execute(
            { objectMapper.writeValueAsString(token) },
            TaskContext.of("RefreshToken", "Serialize", token.refreshTokenId()),
        )
    }

    private fun deserializeToken(json: String): RefreshToken {
        return executor.execute(
            { objectMapper.readValue(json, RefreshToken::class.java) },
            TaskContext.of("RefreshToken", "Deserialize", json.take(30)),
        )
    }
}
